### PR TITLE
core: syncer: add syncer

### PR DESCRIPTION
**Pull request by haoping-ku**
Added module `syncer` for syncing full node.

### CODE REVIEW

File: core/syncer/consensus.go [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+				func() {
+					con.lock.Lock()
+					defer con.lock.Unlock()
+					// If round is cut in agreements, do not add blocks with round less
+					// then cut round.
+					if b.Position.Round < con.agreementRoundCut {
+						return
+					}
+					con.blocks[chainID] = append(con.blocks[chainID], b)
+					sort.Sort(con.blocks[chainID])
```
Comment (jimmyhu-dexon) [inline core/syncer/consensus.go]: `types.ByPosition` implements heap interface. Use heap operation instead of sort.

**Comment by jimmyhu-dexon** [comment]
Please make a separate PR for changes in `core/test` to reduce the size of this one.

File: core/syncer/agreement.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+			}
+		}
+	}
+}
+
+func (a *agreement) handleBlock(b *types.Block) {
+	if _, exist := a.agreementResults[b.Hash]; exist {
+		a.confirm(b)
+	} else {
+		if _, exist := a.blocks[b.Position]; !exist {
```
[the remark this answers is not in the document]
Comment (haoping-ku) [reply core/syncer/agreement.go]: I think there is no need to have sanity check here. Blocks are confirmed after the arrival of verified agreement results.

**Comment by missionliao** [comment]
I'm involving with this PR, it's not appropriate to perform code review against it.

File: core/syncer/agreement.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimmyhu-dexon) [reply core/syncer/agreement.go]: OK

File: core/syncer/consensus.go (hunk 1 above)
Comment (jimmyhu-dexon) [reply core/syncer/consensus.go]: @haoping-ku Please fix this or add a TODO here.

Comment (haoping-ku) [reply core/syncer/consensus.go]: Heap is not sorted. We need a sorted one.

**Pull request merged**